lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectData.java

```diff
@@ -379,12 +379,14 @@ private ClassAccessorData(Class<?> c) {
      * Return the field accessors as an array, indexed by the field index of the
      * given schema.
      */
-    private synchronized FieldAccessor[] getAccessorsFor(Schema schema) {
-      // if synchronized is removed from this method, adjust bySchema appropriately
+    private FieldAccessor[] getAccessorsFor(Schema schema) {
+      // to avoid synchronization, we replace the map for each modification
       FieldAccessor[] result = bySchema.get(schema);
       if (result == null) {
         result = createAccessorsFor(schema);
+        Map<Schema, FieldAccessor[]> bySchema = new WeakHashMap<>(this.bySchema);
         bySchema.put(schema, result);
+        this.bySchema = bySchema;
```

Comment:
Without synchronization `this.bySchema` could be modified several times by now. Would that be a problem ?

Reply:
Since the object is derived It won't cause an issue. There could be duplicated generation across threads. Causing some wasted compute at the start of processing. But since `this.bySchema` is always replaced with a new map with no future modifications concurrent read are safe.

Reply:
The remark about `volatile` is valid: that's exactly what it does.

Also a few notes about `WeakHashMap`:

1. it has weak keys, so it'll only cleanup entries that can no longer be looked up
2. its internal hash table is a singly-linked list, and the update to remove an entry is an atomic operation (that never yields an invalid data structure)
3. the update happens without a memory barrier, which means that the worst case scenario is that the update is not yet available to the other thread, which means that the entry to be removed is read & skipped over in the concurrent call to `get`

In practical terms, this means that because the we update a *copy* of the `WeakHashMap` before it becomes available, and it's only modified to expunge stale entries after it becomes available, the result is effectively thread safe.